### PR TITLE
[sBTC DR] Introduce property-based tests for asset.clar

### PR DESCRIPTION
**Pull request by moodmosaic**
# Introduce property-based tests for asset.clar

Invariant properties allow you to verify that certain conditions always hold true—not only that nothing bad ever happens, but also that something good is always possible. 

This technique became recently popular in the EVM community under the term 'property-based fuzzing' and (to my knowledge) was pioneered by dapp.tools and echidna, and became mainstream by foundry.<sup>1</sup>

This PR contains an encoding of this technique using [fast-check](https://github.com/dubzzz/fast-check) and clarinet, to falsify user-defined invariants (properties) instead of looking just for crashes like a traditional fuzzer. 

```
wallet_3         transfer wallet_7   1140351122 (discarded)
wallet_2 get-total-supply                     0
deployer             mint   faucet   1119048227 bitcoin-tx 0168ee41db8a4766efe02bba1ebc0de320bc1b0abb7304f5f104818a9dd721cf
  faucet get-total-supply            1119048227
deployer             mint wallet_7   2133632941 bitcoin-tx 255dcfd00b04456288b5aacc5275835add15c89067f082952331b7fc1b87a63c
deployer             mint wallet_6   1910196597 bitcoin-tx 255dcfd00b04456288b5aacc5275835add15c89067f082952331b7fc1b87a63c
deployer             mint wallet_4    768982343 bitcoin-tx 255dcfd00b04456288b5aacc5275835add15c89067f082952331b7fc1b87a63c
  faucet         transfer wallet_4   1299487135 (expected, non-owner)
wallet_3      get-balance wallet_6   1910196597
wallet_4         transfer wallet_2   1114063953 (discarded)
wallet_4 get-total-supply            5931860108
wallet_3         transfer wallet_2   1497519370 (discarded)
wallet_4         transfer deployer    169030480
wallet_8 get-total-supply            5931860108
wallet_1 get-total-supply            5931860108
wallet_5 get-total-supply            5931860108
wallet_6 get-total-supply            5931860108
deployer             burn   faucet   1058352425 bitcoin-tx 0168ee41db8a4766efe02bba1ebc0de320bc1b0abb7304f5f104818a9dd721cf
wallet_3         transfer wallet_3    806137080 (discarded)
deployer             mint wallet_8    407684029 bitcoin-tx 255dcfd00b04456288b5aacc5275835add15c89067f082952331b7fc1b87a63c
deployer             burn wallet_6   1382279672 bitcoin-tx 255dcfd00b04456288b5aacc5275835add15c89067f082952331b7fc1b87a63c
wallet_8      get-balance wallet_4    599951863
wallet_8 get-total-supply            3898912040
```

## Summary of Changes

1. Adds `asset_test.ts` and accompanying files.
2. Resolves https://github.com/Trust-Machines/stacks-sbtc/issues/691.
3. Makes sure certain conditions for `asset.clar` always hold true.

### Risks

No risks associated with this PR. Existing tests aren't modified, only new tests added ([append-only tests](https://blog.ploeh.dk/2013/04/02/why-trust-tests/#b23efe23c1574f2b8ff85fb3c529ec9d)).

### How were these changes tested?

Once you `cd` into `romeo/asset-contract` directory, you can run these tests via

```bash
clarinet test tests/asset_test.ts
```

### What future testing should occur?

A few more properties/checks might be added after discussing with @friedger.

For example, we can keep track of how much we minted on each Bitcoin transaction. This may be even added to this PR, before merging it. ✅ Done in https://github.com/stacks-network/sbtc/pull/152/commits/f93e40313b1bc3b7344b1fed04ac341fd47b3387 with the help of @friedger.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

(Let me know if there's a style guideline of this project that I can follow, and any changes to the documentation that I should make. There are no dependent changes in downstream modules.)

---

<sup>1</sup>A more formal description of the technique may be found in [this](https://www.cs.tufts.edu/~nr/cs257/archive/john-hughes/quviq-testing.pdf) paper.

### CODE REVIEW

File: romeo/asset-contract/tests/asset_Commands.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+          r.wallet,
+        )
+      ),
+
+    fc
+      .record({
+        sender: fc.constantFrom(...accounts.values()),
+        amount: fc.integer({ min: 1 }),
+        holder: fc.constantFrom(...accounts.values()),
+        wallet: fc.constantFrom(...accounts.values()),
```
Comment (AshtonStephens) [inline romeo/asset-contract/tests/asset_Commands.ts:123]: Just curiosity:

Are there problems if the accounts sending and holding are the same? If we needed them to be unique but all from the list how would we do that?

Comment (moodmosaic) [reply romeo/asset-contract/tests/asset_Commands.ts:123]: It be checked before running the test: https://github.com/stacks-network/sbtc/pull/152/files#diff-37f7b3538c8e0981f5d429a974ce04c59b31c740ae25826d8da7cf475f39b29aR28

**Review by friedger — APPROVED** [review]
This looks interesting.

Can you add some of the properties that you tested for? Like 
* burn call always succeeds if called by deployer and amount is smaller or equal to the user's balance
* get-balance call always returns the sum of mints, burns and transfers of the user
* get-total-supply call always return the sum of all users wallet's balance
* ...

There is lots of boiler plate code, I am not sure whether I like to write these Commands for my contracts. Could these be annotations on the contract function code?

LGTM!

**Comment by moodmosaic** [comment]
> Can you add some of the properties that you tested for? [..]

* burn call always succeeds if [called by deployer](https://github.com/stacks-network/sbtc/pull/152/files#diff-8068bb48691241676fad63c0c325cf4295211066ce78c0c2813494f75f2c495dR19) and [amount is smaller or equal to the user's balance](https://github.com/stacks-network/sbtc/pull/152/files#diff-ad5217b1068abdeaa8ae7e66d0291ed368b68076faf612e23c5c84c448f6da41R47-R48)
* get-balance call always returns the sum of [mints](https://github.com/stacks-network/sbtc/pull/152/files#diff-ccb366c418f18e6eaabd9f3a59de4418ebf92e6036d1b2805ef0a90273fe4a9aR78), [burns](https://github.com/stacks-network/sbtc/pull/152/files#diff-ad5217b1068abdeaa8ae7e66d0291ed368b68076faf612e23c5c84c448f6da41R82) and [transfers](https://github.com/stacks-network/sbtc/pull/152/files#diff-37f7b3538c8e0981f5d429a974ce04c59b31c740ae25826d8da7cf475f39b29aR57-R64) of the user
* get-total-supply call always return the [sum of all users wallet's balance](https://github.com/stacks-network/sbtc/pull/152/files#diff-c0e6b3be546abdbefae7b9a9a1abc939db1b7efe1208eb499dba0579bbda9399R29-R31)
* mint call always succeeds if [called by deployer](https://github.com/stacks-network/sbtc/pull/152/files#diff-8068bb48691241676fad63c0c325cf4295211066ce78c0c2813494f75f2c495dR76)—(*note*: no limit on the amount minted, theorically if you increase the number of test runs (e.g. 2000000) it might fail if exceeds the 21 M sBTC supply = 2.1 Q sats total)
* transfer call always succeeds if [sender is not the recipient wallet and sender has enough funds](https://github.com/stacks-network/sbtc/pull/152/files#diff-37f7b3538c8e0981f5d429a974ce04c59b31c740ae25826d8da7cf475f39b29aR24-R38)
* transfer call always fails [if sender is not the owner and sender has not enough funds](https://github.com/stacks-network/sbtc/pull/152/files#diff-17218601587da61dd73801757b3e7d14b6973378f597fe269c94a783271b6805R48)

**Comment by moodmosaic** [comment]
>There is lots of boiler plate code, I am not sure whether I like to write these Commands for my contracts.

Boilerplate may be removed or minimized using convention of configuration, or even a small DSL (e.g. annotations on the contract function code, as you wrote). [Doable.](https://github.com/stacks-network/sbtc/pull/152/files#diff-8068bb48691241676fad63c0c325cf4295211066ce78c0c2813494f75f2c495dR142)

I want to introduce this particular style of property-based testing by first making everything explicit, and then start minimizing the boilerplate, otherwise (IME) it can be harder for the PR reviewer(s) to follow.

File: romeo/asset-contract/tests/asset_Commands.ts (hunk 1 above)
Comment (AshtonStephens) [reply romeo/asset-contract/tests/asset_Commands.ts:123]: So then we'd still be setting up the test incorrectly but then nixing it before it runs?

**Review by friedger — APPROVED** [review]
LGTM

**Review by stjepangolemac — APPROVED** [review]
Don't know lot about this but it sounds reasonable.

**Pull request merged**